services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/PinnedFlowSpec.groovy

```diff
@@ -0,0 +1,111 @@
+import static org.openkilda.model.MeterId.MAX_SYSTEM_RULE_METER_ID
+import static org.openkilda.testing.Constants.WAIT_OFFSET
+
+import org.openkilda.functionaltests.BaseSpecification
+import org.openkilda.functionaltests.helpers.Wrappers
+import org.openkilda.messaging.info.event.IslChangeType
+import org.openkilda.messaging.payload.flow.FlowState
+import org.openkilda.model.Cookie
+
+class PinnedFlowSpec extends BaseSpecification {
+    def "System doesn't reroute(automatically) pinned flow"() {
+        given: "A pinned flow with alt path available"
+        def switchPair = topologyHelper.getAllNeighboringSwitchPairs().find { it.paths.size() > 1 } ?:
+                assumeTrue("No suiting switches found", false)
+        def flow = flowHelper.randomFlow(switchPair)
+        flow.pinned = true
+        flowHelper.addFlow(flow)
+
+        def currentPath = pathHelper.convert(northbound.getFlowPath(flow.id))
+        def altPath = switchPair.paths.find { it != currentPath }
+        def involvedSwitches = pathHelper.getInvolvedSwitches(flow.id)
+
+        when: "Init reroute by bringing current path's ISL down"
+        def currentIsls = pathHelper.getInvolvedIsls(currentPath)
+        def newIsls = pathHelper.getInvolvedIsls(altPath)
+        def islToBreak = currentIsls.find { !newIsls.contains(it) }
+
+        def cookiesMap = involvedSwitches.collectEntries { sw ->
+            [sw.dpId, northbound.getSwitchRules(sw.dpId).flowEntries.findAll {
+                !Cookie.isDefaultRule(it.cookie)
+            }*.cookie]
+        }
+        def metersMap = involvedSwitches.collectEntries { sw ->
+            [sw.dpId, northbound.getAllMeters(sw.dpId).meterEntries.findAll {
+                it.meterId > MAX_SYSTEM_RULE_METER_ID
+            }*.meterId]
+        }
+
+        northbound.portDown(islToBreak.srcSwitch.dpId, islToBreak.srcPort)
+
+        then: "Flow is not rerouted and marked as DOWN"
+        Wrappers.wait(WAIT_OFFSET) {
+            assert northbound.getFlowStatus(flow.id).status == FlowState.DOWN
+            assert pathHelper.convert(northbound.getFlowPath(flow.id)) == currentPath
+        }
+
+        and: "Rules and meters are not changed"
+        def cookiesMapAfterReroute = involvedSwitches.collectEntries { sw ->
+            [sw.dpId, northbound.getSwitchRules(sw.dpId).flowEntries.findAll {
+                !Cookie.isDefaultRule(it.cookie)
+            }*.cookie]
+        }
+        def metersMapAfterReroute = involvedSwitches.collectEntries { sw ->
+            [sw.dpId, northbound.getAllMeters(sw.dpId).meterEntries.findAll {
+                it.meterId > MAX_SYSTEM_RULE_METER_ID
+            }*.meterId]
+        }
+
+        cookiesMap.sort() == cookiesMapAfterReroute.sort()
+        metersMap.sort() == metersMapAfterReroute.sort()
+
+        when: "The broken ISL is restored"
+        northbound.portUp(islToBreak.srcSwitch.dpId, islToBreak.srcPort)
+
+        then: "The flow is marked as UP and not rerouted"
```

Comment:
I think the alt path should be made more preferable than the current path in order to properly verify this

Reply:
added